### DISCUSSION TITLE
Experimental EV3dev Debian 12 Bookworm, 11 Bullseye, 10 Buster, 9 Stretch images

### DISCUSSION THREAD

**Discussion by Growflavor** [ev3dev/ev3dev#1636] [Ideas]
(WiP) Place to broaden the discussion started at #1554 

& RE: "There is a desire to test approximately similar builds on Bullseye and Stretch" #1558 https://github.com/ev3dev/ev3dev/issues/1558#issuecomment-2173970590

**Debian 12 Bookworm:**

- on the 4.19.85 Kernel:
- https://github.com/ev3dev/ev3dev/issues/1554#issuecomment-2099746486
- boots but networking connections are difficult/intermittent
- processes tend to get killed (OOM?)
- 'tinkerable' but not currently 'useable'
- docker pull growflavor/ev3images:ev3dev12rpycbtbm3
- 
- likely needs a more current kernel (requires people with skill with ARMEL kernels, EV3 drivers, & time):
- https://github.com/ev3dev/ev3-kernel/tree/ev3dev-bookworm
- Advantage would be that Python 3.11 runs remarkably faster on the EV3: see chart at https://github.com/ev3dev/ev3dev/issues/1635#issuecomment-2196269757

**Debian 11 Bullseye:**

- on the 4.19.85 Kernel:
- https://github.com/ev3dev/ev3dev/issues/1554#issuecomment-2019205120
- appears useable for EV3dev
- precompiled pybricks-micropython packages NO longer work -- Bullseye uses several newer versions of the dependency libraries someone with programming experience could assess if a recompile using the newer libraries would resolve
- docker pull growflavor/ev3images:latest

**Debian 10 Buster:**

- on the 4.19.85 Kernel:
- appears very useful including able to build python 3.11, see discussion #1634 for details:
- https://github.com/orgs/ev3dev/discussions/1634
- docker pull growflavor/ev3images:ev3dev10imgv02b

**Debian 9 Stretch:**

- packages can be updated from the 'officially released version'
- here is an example, (note: I think this version had all available backports apt installed which could **potentially** lead to stability issues)
- https://github.com/ev3dev/ev3dev/issues/1554#issuecomment-2087973731

**Comment by Project516** [comment]
Hello, I am interested in helping with bookworm. How would I use the ev3dev kernel bookworm branch to build an image? I know brickstrap creates an image from a docker container but I am not sure how to use the repositories to make one. Could someone could give me an overview for manually compiling everything? Thank you!

**Reply by dlech** [reply]
Debian trixie will be promoted to stable in the next month or so, so if we want to make a serious effort to make a usable ev3dev release, it would probably make sense to target that so that it lasts the longest without getting "too old".

The way the process works is that for any of the package on `archive.ev3dev.org`, we will need to migrate them to trixie. Then we can add a new trixie target to https://github.com/ev3dev/docker-library/. That repo is used to create the docker images that brickstrap uses to create a bootable image.

**Reply by Project516** [reply]
How do I get started? Sorry, I have no experience with this. Could you give me some resources to look at too? Thanks!

**Reply by dlech** [reply]
The kernel is the first place to start since everything else depends on that.

* How we build the kernel: https://github.com/ev3dev/ev3dev-buildscripts
* The kernel source code: https://github.com/ev3dev/ev3-kernel/
* Ev3dev custom drivers: https://github.com/ev3dev/lego-linux-drivers

We need to update this to the v6.12 kernel. About a year ago, I started doing this for the 6.9 kernel in the [ev3dev-bookworm branch](https://github.com/ev3dev/ev3-kernel/commits/ev3dev-bookworm/) but didn't finish.

**Reply by Project516** [reply]
I have started using kernel 6.12.38 as that’s what’s shipped in trixie. The Lego-Linux-drivers need to be modified for the newer kernel.

Also Debian trixie will be the last Debian version for armel.

**Reply by dlech** [reply]
> Also Debian trixie will be the last Debian version for armel.

Yeah, I saw that today. All the more reason to use that version.

**Reply by Project516** [reply]
> The kernel is the first place to start since everything else depends on that.
> 
> * How we build the kernel: https://github.com/ev3dev/ev3dev-buildscripts
> * The kernel source code: https://github.com/ev3dev/ev3-kernel/
> * Ev3dev custom drivers: https://github.com/ev3dev/lego-linux-drivers
> 
> We need to update this to the v6.12 kernel. About a year ago, I started doing this for the 6.9 kernel in the [ev3dev-bookworm branch](https://github.com/ev3dev/ev3-kernel/commits/ev3dev-bookworm/) but didn't finish.

@dlech I have gotten the ev3dev kernel to compile with 6.12 - I rebased the bookworm branch to 6.12, and got it to compile. I have created forks of ev3-kernel (https://github.com/Project516/ev3-kernel), lego-linux-drivers (https://github.com/Project516/lego-linux-drivers) and ev3dev-buildscripts (https://github.com/Project516/ev3dev-buildscripts) with my changes.

**Reply by dlech** [reply]
That's great. If you are interested, you could make some pull requests and I might have some time to have a closer look over the holidays.

**Reply by Project516** [reply]
> That's great. If you are interested, you could make some pull requests and I might have some time to have a closer look over the holidays.

I made a few prs. I wasn't sure what branch to target, so I targeted the closest one to trixie. I have gotten the ev3-systemd and brickman (and the grx and ev3devkit dependencies) to compile (though not tested), so hopefully I can work on getting a testable image that boots into brickman.

**Reply by Muresan73** [reply]
I built a kernel successfully from your branch @Project516 ( i had to manually set `LOADADDR=0xC0008000` in local-env )
 but i was not able to successfully get it to boot.  With your test img before systemd could start it hangs. 
 
 Then i tried with the last "offical" release ev3dev-stretch-ev3-generic-2020-04-10 image with boot.scr and u-boot.bin in boot partition but now it does not even load the new kernel uImage :/ ( it boots normally to v4.14),  i tried the Custom Locations trick also.
 
What is the right approach? Do you guys have some advice?
 I would love to help out with the testing and debugging if needed :)

**Reply by dlech** [reply]
When I'm working on the kernel, I usually edit the `uEnv.txt` to boot from a `uImage` and `.dtb` in the BOOT partition. This way, if I break something, it can easily be fixed by just deleting those files.

It has been quite a while since I did this though, so I don't remember the changes I made to `uEnv.txt` off of the top of my head.

**Reply by Muresan73** [reply]
The problem was that i used a preexisting boot.scr ( the build-boot-src mentioned in the readme does not exist anymore) which pointed to the wrong location.
I managed to make it work by adding these to my `uEnv.txt`
```
mmcdev=0         # select sd drive
bootpart=0:1     # select BOOT partition which is the first one
bootdir=/        # path to the boot folder
bootfile=uImage  # name of the image file
fdtfile=da850-lego-ev3.dtb
```

**Reply by Project516** [reply]
@Muresan73 Did you get the 6.12 kernel to boot with the stretch image? What steps did you take? I haven’t tested the kernel yet.

**Reply by Muresan73** [reply]
I needed to add the bootloader from the stretch image to the boot partition ( maybe i should just flash a newer uboot to the ev3 at this point ) 

I got the errors in the boot process:
```
ev3-board board: Failed to get hw-id nvmem -2
ev3-board board: probe with driver ev3-board failed with error -2
```

I fixed it with updating the dts file for the eeprom, now it just hangs with no specific error.

If i manage to get somewhere more tangible i will add more details.

**Reply by Muresan73** [reply]
Small Progress Update:
Successfully built a [Linux 6.18.3](https://github.com/Project516/ev3-kernel/commit/b0d7fd485c7bb6f0aa3df1b980b6c014f622bdd2) kernel image using Buildroot for the EV3. Some LEGO drivers were causing the boot process to hang, so I created a minimal device tree with no lego peripherals  enabled. This minimal configuration boots perfectly, this confirms that the basic boot chain works!
